src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlerTests.cs

```diff
@@ -536,5 +532,170 @@ public static void ArrayPreserveDuplicateDictionaries()
             Assert.Same(rootCopy[0], rootCopy[1]);
         }
         #endregion
+
+        #region ReferenceResolver
+        [Fact]
+        public static void CustomReferenceResolver()
+        {
+            string json = @"[
+  {
+    ""$id"": ""0b64ffdf-d155-44ad-9689-58d9adb137f3"",
+    ""Name"": ""John Smith"",
+    ""Spouse"": {
+      ""$id"": ""ae3c399c-058d-431d-91b0-a36c266441b9"",
+      ""Name"": ""Jane Smith"",
+      ""Spouse"": {
+        ""$ref"": ""0b64ffdf-d155-44ad-9689-58d9adb137f3""
+      }
+    }
+  },
+  {
+    ""$ref"": ""ae3c399c-058d-431d-91b0-a36c266441b9""
+  }
+]";
+            var options = new JsonSerializerOptions
+            {
+                WriteIndented = true,
+                ReferenceHandler = new ReferenceHandler<GuidReferenceResolver>()
+            };
+            ImmutableArray<PersonReference> people = JsonSerializer.Deserialize<ImmutableArray<PersonReference>>(json, options);
+
+            Assert.Equal(2, people.Length);
+
+            PersonReference john = people[0];
+            PersonReference jane = people[1];
+
+            Assert.Same(john, jane.Spouse);
+            Assert.Same(jane, john.Spouse);
+
+            Assert.Equal(json, JsonSerializer.Serialize(people, options));
+        }
+
+        [Fact]
+        public static void CustomReferenceResolverPersistent()
+        {
+            var options = new JsonSerializerOptions
+            {
+                WriteIndented = true,
+                ReferenceHandler = new PresistentGuidReferenceHandler
+                {
+                    PersistentResolver = new GuidReferenceResolver()
```

Comment:
Are there tests that re-use the same resolver instance across multiple (de)serialization requests? (one of the scenarios discussed)